## PPAC traking
利用多个PPAC的x，y信息进行traking的原理如图所示,图中红点代表探测器测量的位置。由于PPAC的探测效率，对于每个入射粒子只有部分PPAC的x或y方向可给出位置信息。假设对于某一个入射粒子，有两个以上(含两个)探测器给出x位置信息 (如图:1Ax,1Bx,3x) 时，入射粒子的x-z径迹可由上述x-z点进行线性拟合得到。y-z径迹的处理方法与x-z径迹的处理方法一致。利用上述线性方程，可内插或外推某一z位置的x和y信息。图中空心黑点代表由线性方程计算得到的位置。
 ![setup](traking.png) 
 
 #### 束流径迹
 ![setup](pos_trak.png)  
 
 下面代码假设同时有1A，2A，3探测器给出x，y的位置信息。

### 利用makeclass生成 tracking.h, tracking.C
``` cpp
root -l run00005.root
>tree->MakeClass("tracking");
>.q
编辑 tracking.h和tracking.C 保存
使用方法：
root -l
> .L tracking.C
> tracking t
>t.Loop()
>.q

root -l tracking.root //分析
```

### tracking.h
<font color="Red">在traking.h 代码中增加了用户变量，成员函数SetBranch()以及TrackInit() 定义</font> 

``` cpp
//////////////////////////////////////////////////////////
// This class has been automatically generated on
// Thu Mar  7 11:27:23 2019 by ROOT version 6.14/02
// from TTree tree/tree must2
// found on file: run0005.root
//////////////////////////////////////////////////////////

#ifndef tracking_h
#define tracking_h

#include <TROOT.h>
#include <TChain.h>
#include <TFile.h>

// Header file for the classes stored in the TTree if any.

class tracking {
public :
   TTree          *fChain;   //!pointer to the analyzed TTree or TChain
   Int_t           fCurrent; //!current Tree number in a TChain

// Fixed size dimensions of array or collections stored in the TTree if any.

   // Declaration of leaf types
   Int_t           RunNumber;
   Int_t           EventNumber;
   Int_t           beamTrig;
   Int_t           must2Trig[8];
   Float_t         PPACF5[5][5];
   Float_t         PPACF8[5][5];
   Float_t         PPACF9[5][5];
   Float_t         FPTime[12][6];
   Float_t         FPdE[12][3];
   Float_t         FPPosition[12][4];
   Float_t         F5PPACRawData[5][10];
   Float_t         F8PPACRawData[5][10];
   Float_t         F9PPACRawData[5][10];
   Float_t         F10PPACRawData[5][10];
   Float_t         F11PPACRawData[5][10];
   Float_t         F3PlaRawData[4];
   Float_t         F7PlaRawData[4];
   Float_t         F11PlaRawData[4];
   Float_t         F11ICRawData[10];
   Float_t         TOF[8];
   Float_t         AoQ[9][4];

  //by user
   Double_t xx[3],xz[3],yy[3],yz[3];//1A,2A,3
   Double_t p2bx,p2by;//2B x-y
   Double_t dx[3],dy[3];
   Double_t tx,ty;//target pos
   Double_t c2nx,c2ny;//chi2/ndf for xfit,yfit
  
   
   // List of branches
   TBranch        *b_RunNumber;   //!
   TBranch        *b_EventNumber;   //!
   TBranch        *b_beamTrig;   //!
   TBranch        *b_must2Trig;   //!
   TBranch        *b_PPACF5;   //!
   TBranch        *b_PPACF8;   //!
   TBranch        *b_PPACF9;   //!
   TBranch        *b_FPTime;   //!
   TBranch        *b_FPdE;   //!
   TBranch        *b_FPPosition;   //!
   TBranch        *b_F5PPACRawData;   //!
   TBranch        *b_F8PPACRawData;   //!
   TBranch        *b_F9PPACRawData;   //!
   TBranch        *b_F10PPACRawData;   //!
   TBranch        *b_F11PPACRawData;   //!
   TBranch        *b_F3PlaRawData;   //!
   TBranch        *b_F7PlaRawData;   //!
   TBranch        *b_F11PlaRawData;   //!
   TBranch        *b_F11ICRawData;   //!
   TBranch        *b_TOF;   //!
   TBranch        *b_AoQ;   //!

   tracking(TTree *tree=0);
   virtual ~tracking();
   virtual Int_t    Cut(Long64_t entry);
   virtual Int_t    GetEntry(Long64_t entry);
   virtual Long64_t LoadTree(Long64_t entry);
   virtual void     Init(TTree *tree);
   virtual void     Loop();//by user
   virtual void     SetBranch(TTree *tree);//by user
   virtual void     TrackInit();//by user
   virtual Bool_t   Notify();
   virtual void     Show(Long64_t entry = -1);

};

#endif

#ifdef tracking_cxx
tracking::tracking(TTree *tree) : fChain(0) 
{
// if parameter tree is not specified (or zero), connect the file
// used to generate this class and read the Tree.
   if (tree == 0) {
      TFile *f = (TFile*)gROOT->GetListOfFiles()->FindObject("run0005.root");
      if (!f || !f->IsOpen()) {
         f = new TFile("run0005.root");
      }
      f->GetObject("tree",tree);

   }
   Init(tree);
}

tracking::~tracking()
{
   if (!fChain) return;
   delete fChain->GetCurrentFile();
}

Int_t tracking::GetEntry(Long64_t entry)
{
// Read contents of entry.
   if (!fChain) return 0;
   return fChain->GetEntry(entry);
}
Long64_t tracking::LoadTree(Long64_t entry)
{
// Set the environment to read one entry
   if (!fChain) return -5;
   Long64_t centry = fChain->LoadTree(entry);
   if (centry < 0) return centry;
   if (fChain->GetTreeNumber() != fCurrent) {
      fCurrent = fChain->GetTreeNumber();
      Notify();
   }
   return centry;
}

void tracking::Init(TTree *tree)
{
   // The Init() function is called when the selector needs to initialize
   // a new tree or chain. Typically here the branch addresses and branch
   // pointers of the tree will be set.
   // It is normally not necessary to make changes to the generated
   // code, but the routine can be extended by the user if needed.
   // Init() will be called many times when running on PROOF
   // (once per file to be processed).

   // Set branch addresses and branch pointers
   if (!tree) return;
   fChain = tree;
   fCurrent = -1;
   fChain->SetMakeClass(1);

   fChain->SetBranchAddress("RunNumber", &RunNumber, &b_RunNumber);
   fChain->SetBranchAddress("EventNumber", &EventNumber, &b_EventNumber);
   fChain->SetBranchAddress("beamTrig", &beamTrig, &b_beamTrig);
   fChain->SetBranchAddress("must2Trig", must2Trig, &b_must2Trig);
   fChain->SetBranchAddress("PPACF5", PPACF5, &b_PPACF5);
   fChain->SetBranchAddress("PPACF8", PPACF8, &b_PPACF8);
   fChain->SetBranchAddress("PPACF9", PPACF9, &b_PPACF9);
   fChain->SetBranchAddress("FPTime", FPTime, &b_FPTime);
   fChain->SetBranchAddress("FPdE", FPdE, &b_FPdE);
   fChain->SetBranchAddress("FPPosition", FPPosition, &b_FPPosition);
   fChain->SetBranchAddress("F5PPACRawData", F5PPACRawData, &b_F5PPACRawData);
   fChain->SetBranchAddress("F8PPACRawData", F8PPACRawData, &b_F8PPACRawData);
   fChain->SetBranchAddress("F9PPACRawData", F9PPACRawData, &b_F9PPACRawData);
   fChain->SetBranchAddress("F10PPACRawData", F10PPACRawData, &b_F10PPACRawData);
   fChain->SetBranchAddress("F11PPACRawData", F11PPACRawData, &b_F11PPACRawData);
   fChain->SetBranchAddress("F3PlaRawData", F3PlaRawData, &b_F3PlaRawData);
   fChain->SetBranchAddress("F7PlaRawData", F7PlaRawData, &b_F7PlaRawData);
   fChain->SetBranchAddress("F11PlaRawData", F11PlaRawData, &b_F11PlaRawData);
   fChain->SetBranchAddress("F11ICRawData", F11ICRawData, &b_F11ICRawData);
   fChain->SetBranchAddress("TOF", TOF, &b_TOF);
   fChain->SetBranchAddress("AoQ", AoQ, &b_AoQ);
   Notify();
}

Bool_t tracking::Notify()
{
   // The Notify() function is called when a new file is opened. This
   // can be either for a new TTree in a TChain or when when a new TTree
   // is started when using PROOF. It is normally not necessary to make changes
   // to the generated code, but the routine can be extended by the
   // user if needed. The return value is currently not used.

   return kTRUE;
}

void tracking::Show(Long64_t entry)
{
// Print contents of entry.
// If entry is not specified, print current entry
   if (!fChain) return;
   fChain->Show(entry);
}
Int_t tracking::Cut(Long64_t entry)
{
// This function may be called from Loop.
// returns  1 if entry is accepted.
// returns -1 otherwise.
   return 1;
}
#endif // #ifdef tracking_cxx


```

### tracking.C
<font color="Red">在tracking的基础上进行了修改</font> 
``` cpp
#define tracking_cxx
#include "tracking1.h"
#include <TH2.h>
#include <TStyle.h>
#include <TCanvas.h>
#include <TF1.h>
#include <TFitResult.h>

void tracking::SetBranch(TTree *tree)
{
  //measured pos
  tree->Branch("xx",&xx,"xx[3]/D");//1A,2A,3
  tree->Branch("xz",&xz,"xz[3]/D");
  tree->Branch("yy",&yy,"yy[3]/D");
  tree->Branch("yz",&yz,"yz[3]/D");

  //difference between measured and calculated -for pos resolution.
  tree->Branch("dx",&dx,"dx[3]/D");
  tree->Branch("dy",&dy,"dy[3]/D");

  //for efficiency calculation -2B
  tree->Branch("p2bx",&p2bx,"p2bx/D");
  tree->Branch("p2by",&p2by,"p2by/D");
  
  //target x-y
  tree->Branch("tx",&tx,"tx/D");
  tree->Branch("ty",&ty,"ty/D");
  
  //ch2/ndf for linear fitting.
  tree->Branch("c2nx",&c2nx,"c2nx/D");
  tree->Branch("c2ny",&c2ny,"c2ny/D");
}

void tracking::TrackInit()
{
  tx=-999;
  ty=-999;
  
  //1A
  xx[0]=PPACF8[0][0];
  yy[0]=PPACF8[0][1];
  xz[0]=PPACF8[0][2];
  yz[0]=PPACF8[0][3];
 
  //2A
  xx[1]=PPACF8[2][0];
  yy[1]=PPACF8[2][1];
  xz[1]=PPACF8[2][2];
  yz[1]=PPACF8[2][3];
   
  //3
  xx[2]=PPACF8[4][0];
  yy[2]=PPACF8[4][1];
  xz[2]=PPACF8[4][2];
  yz[2]=PPACF8[4][3];

  //2B
  p2bx=PPACF8[3][0];
  p2by=PPACF8[3][1];
 
}
void tracking::Loop()
{
  TFile *opf=new TFile("tracking.root","recreate");
  TTree *tree=new TTree("tree","ppac traking");
  SetBranch(tree);
  
   if (fChain == 0) return;
   Long64_t nentries = fChain->GetEntriesFast();
   Long64_t nbytes = 0, nb = 0;
   for (Long64_t jentry=0; jentry<nentries;jentry++) {
      Long64_t ientry = LoadTree(jentry);
      if (ientry < 0) break;
      nb = fChain->GetEntry(jentry);   nbytes += nb;
      
      TrackInit();
      bool b1a=xx[0]>-999 && yy[0]>-999;
      bool b2a=xx[1]>-999 && yy[1]>-999;
      bool b3=xx[2]>-999 && yy[2]>-999;
      if(!b1a || !b2a || !b3) continue;
      
      //fit x-z trajectory
      TFitResultPtr r;
      TGraph *grx=new TGraph(3,xz,xx);
      TF1 *fx=new TF1("fx","pol1",-2000,0);
      r=grx->Fit(fx,"SQ");
      tx=fx->Eval(0);
      for(int i=0;i<3;i++) dx[i]=xx[i]-fx->Eval(xz[i]);
      c2nx=r->Chi2()/r->Ndf();
      delete grx;
      delete fx;

      //fit y-z trajectory      
      TGraph *gry=new TGraph(3,yz,yy);
      TF1 *fy=new TF1("fy","pol1",-2000,0);
      r=gry->Fit(fy,"SQ");      
      ty=fy->Eval(0);
      for(int i=0;i<3;i++) dy[i]=yy[i]-fy->Eval(yz[i]);
      c2ny=r->Chi2()/r->Ndf();
      delete gry;
      delete fy;

      tree->Fill();
      if(jentry%10000==0) cout<<"processing "<<jentry<<endl;
      
   }
   tree->Write();
   opf->Close();
}

```

In [1]:
%jsroot on
TFile *ipf=new TFile("tracking.root");
TTree *tree=(TTree*) ipf->Get("tree");
TCanvas *c1=new TCanvas("c1","c1");
tree->Print();

******************************************************************************
*Tree    :tree      : ppac traking                                           *
*Entries :   169686 : Total =        32675458 bytes  File  Size =   21057944 *
*        :          : Tree compression factor =   1.55                       *
******************************************************************************
*Br    0 :xx        : xx[3]/D                                                *
*Entries :   169686 : Total  Size=    4084163 bytes  File Size  =    2817086 *
*Baskets :      128 : Basket Size=      32000 bytes  Compression=   1.45     *
*............................................................................*
*Br    1 :xz        : xz[3]/D                                                *
*Entries :   169686 : Total  Size=    4084163 bytes  File Size  =      30796 *
*Baskets :      128 : Basket Size=      32000 bytes  Compression= 132.53     *
*...................................................

In [2]:
tree->Draw("ty:tx>>htxy(500,-100,100,500,-100,100)","","colz");
c1->Draw();//从图像上可看出由于shilding造成的方块区域。

$\chi^2/Ndf :  tx$

In [3]:
tree->Draw("c2nx:tx>>(400,-100,100,200,0,1000)","","colz");
c1->Draw();//从chi2/ndf图上可看出，部分事件的径迹拟合误差很大，这部分要在后续数据处理中去掉。

In [4]:
tree->Draw("dx[1]>>(200,-5,5)");
c1->Draw();//图中dx[1]的分布的宽度，对应于相应探测器的位置分辨率。

In [5]:
tree->Draw("p2bx:p2by>>(200,-100,100,200,-100,100)","p2bx>-999 && p2by>-999","colz");
c1->Draw();

对比上两张图的计数(dx[1]，没有卡条件，计数对应三个探测器都有x，y位置的数目。上一张图计数对应2bx-y都有信号的数目)，可得2B探测器的位置探测效率。PPAC82B的位置探测效率(x-y)为 $\epsilon_{xy}= \frac{147377}{169686}=86.8\%$.